### PR TITLE
lint: wip

### PR DESCRIPTION
**Pull request by mjibson**
https://github.com/mjibson/materialize/pull/new/lint

---
This change is [Reviewable](https://reviewable.io/reviews/materializeinc/materialize/4300)

### CODE REVIEW

File: Cargo.toml [hunk 1]
```diff
@@ -14,6 +14,7 @@ members = [
     "src/expr",
     "src/interchange",
     "src/kafka-util",
+    "src/lint",
```
Comment (benesch) [inline Cargo.toml:17]: A slightly more specific name than lint would be ideal! Like "sql-lint" or "expr-lint" or somesuch.

**Review by benesch — COMMENTED** [review]
Seems sensible! There is a bit of a policy question here though of whether these lints should run before or after optimization. As you've written them pre-optimization they're going to fire on stuff like

```sql
SELECT * FROM a CROSS JOIN b WHERE a.col = b.col
```

even if the optimizer is successfully able to identify the join condition for `a, b` and therefore do something smarter than a cross join. (This actually comes up a lot with pre SQL 92 joins, like `SELECT * FROM a, b ...`, where you _had_ to do a cross join and filter it down with a `WHERE`.)

Perhaps @justinj and @frankmcsherry have thoughts on that point?

**Comment by frankmcsherry** [comment]
Some lints should certainly run after optimization, especially if we want to warn people about the potential execution. Perhaps there are lints that should run before optimization, for example warning people about implicit casts or other potentially defective SQLisms.

**Comment by justinj** [comment]
Agreed on the CROSS JOIN thing, and I'm not sure how I feel about warning about the UNION/OUTER JOIN ones, since those might be intentional? My expectation around a lint rule is that everything it tells me should be actionable (except maybe in exceptional cases). I can't think of something obviously lintable off the top of my head for Materialize, but an obvious example would be in OLTP databases, where `OFFSET` is basically always the wrong thing to use.

**Comment by mjibson** [comment]
I was under the impression that keeping around the map from SQL to post-optimization plan is currently difficult. If that's the case, is there a better way to solve this problem well enough? Like have a thing that analyzes final optimized plans looking for bad stuff and says like "found an <X> in the dataflow plan that's bad, which might have been produced by a SQL <Y>, consider changing <Y> to <Z> to produce a better plan". This will avoid the work to track the SQL but still identify potentially bad stuff?

**Comment by frankmcsherry** [comment]
My take, though admittedly not as a SQL person, is that linting the query plan seems pretty good and makes sense. It is not always immediately actionable (e.g. "you have a cross join!", "no I don't, I have a join that equate expressions rather than columns"), but it does seem to warn about the right thing (the plan we are about to run). Random thoughts though, not any clear definition on my part!

**Comment by justinj** [comment]
I think linting the plan is good too! I think hoping to insulate users from inspecting the plans at all is desirable but I also think the prior art here for most databases is that they expect users to be able to interpret plans themselves (happy to be shown evidence that's not true, though!), so I think the place we should invest effort on this front is documentation on how users can understand what plans are doing (for instance, [this post](https://www.cockroachlabs.com/blog/index-selection-cockroachdb-2/) leveled me up substantially).

**Pull request closed without merging**